lgc/test/Transforms/CpsLowering/cps-stack-lowering.lgc

```diff
@@ -0,0 +1,234 @@
+; RUN: lgc -mcpu=gfx1030 -o - -passes='require<lgc-pipeline-state>,lgc-patch-entry-point-mutate' %s | FileCheck --check-prefixes=CHECK %s
+
+declare void @lgc.cps.jump(...) noreturn
+declare ptr addrspace(32) @lgc.cps.alloc(i32)
+declare void @lgc.cps.free(i32)
+declare i32 @lgc.cps.as.continuation.reference(ptr)
+declare ptr addrspace(32) @lgc.cps.peek(i32)
+
+define void @test.0({} %unused) !lgc.cps !{i32 1} {
+  %mem = call ptr addrspace(32) @lgc.cps.alloc(i32 10)   ; round up to 12 during lowering
+
+  store i32 333, ptr addrspace(32) %mem
+
+  %p1 = getelementptr i32, ptr addrspace(32) %mem, i32 1
+  store i32 111, ptr addrspace(32) %p1
+
+  %p2 = getelementptr i8, ptr addrspace(32) %mem, i32 9
+  store i8 99, ptr addrspace(32) %p2
+
+  %q1 = ptrtoint ptr addrspace(32) %p1 to i32
+
+  %state = insertvalue { ptr addrspace(32) } poison, ptr addrspace(32) %p2, 0
+
+  %cr = call i32 @lgc.cps.as.continuation.reference(ptr @test.1)
+  call void (...) @lgc.cps.jump(i32 %cr, i32 2, { ptr addrspace(32) } %state, ptr addrspace(32) %p2, i32 %q1)
+  unreachable
+}
+
+define void @test.1({} %no_state, ptr addrspace(32) %p2, i32 %q1) !lgc.cps !{i32 1} {
+  %p1 = inttoptr i32 %q1 to ptr addrspace(32)
+  %n111 = load i32, ptr addrspace(32) %p1
+  %n99 = load i8, ptr addrspace(32) %p2
+
+  %cr = call i32 @lgc.cps.as.continuation.reference(ptr @test.2)
+  call void (...) @lgc.cps.jump(i32 %cr, i32 2, {} poison)
+  unreachable
+}
+
+define void @test.2({ ptr addrspace(32) } %state) !lgc.cps !{i32 1} {
+  %mem = call ptr addrspace(32) @lgc.cps.peek(i32 10)    ; round up to 12 during lowering
+  %p2 = extractvalue { ptr addrspace(32) } %state, 0
+
+  %n333 = load i32, ptr addrspace(32) %mem
+  %n99 = load i8, ptr addrspace(32) %p2
+
+  call void @lgc.cps.free(i32 10)   ; round up to 12 during lowering
+  ret void
+}
+; CHECK-LABEL: define {{[^@]+}}@test.0
+; CHECK-SAME: (i32 inreg [[GLOBALTABLE:%.*]], i32 inreg [[PERSHADERTABLE:%.*]], ptr addrspace(4) inreg [[NUMWORKGROUPSPTR:%.*]], i32 inreg [[TMP0:%.*]], i32 inreg [[TMP1:%.*]], i32 inreg [[TMP2:%.*]], i32 inreg [[TMP3:%.*]], i32 inreg [[TMP4:%.*]], i32 inreg [[TMP5:%.*]], i32 inreg [[TMP6:%.*]], i32 inreg [[TMP7:%.*]], i32 inreg [[TMP8:%.*]], i32 inreg [[TMP9:%.*]], i32 inreg [[TMP10:%.*]], i32 inreg [[SPILLTABLE:%.*]], i32 [[VCR:%.*]], ptr addrspace(5) [[VSP:%.*]]) #[[ATTR1:[0-9]+]] align 64 !lgc.cps !2 {
+; CHECK-NEXT:    [[TMP12:%.*]] = alloca ptr addrspace(5), align 4, addrspace(5)
+; CHECK-NEXT:    [[TMP13:%.*]] = call i64 @llvm.amdgcn.s.getpc()
+; CHECK-NEXT:    [[TMP14:%.*]] = bitcast i64 [[TMP13]] to <2 x i32>
+; CHECK-NEXT:    store ptr addrspace(5) [[VSP]], ptr addrspace(5) [[TMP12]], align 4
+; CHECK-NEXT:    [[TMP15:%.*]] = load ptr addrspace(5), ptr addrspace(5) [[TMP12]], align 4
+; CHECK-NEXT:    [[TMP16:%.*]] = getelementptr ptr addrspace(5), ptr addrspace(5) [[TMP15]], i32 12
```

Comment:
The GEP type here needs to be `i8` to use byte offsetting.

Reply:
This is a really nice catch, Thanks! I have fixed all these issues.